tutorial: add second part of tutorial series for building DReddit (#135)
embark-framework/embark-site: source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-2.md

title: Building a decentralized Reddit with Embark - Part 2
summary: "This is the second part of the three part tutorial about building a decentralized Reddit with Embark. In this part, we'll be focussing on testing our Smart Contract using EmbarkJS."
categories:
  - tutorials
---

In [the first part of this tutorial](/news/2019/01/28/building-a-decentralized-reddit-with-embark-part-1/) we've implemented a `DReddit` Smart Contract that comes with methods to create and vote on topic posts. In this part we'll continue right where we've left off and take a closer look at how we can test our Smart Contract using Embark.

And off we go!

## Writing a first test

We've got plenty functionality to cover in our tests, but let's start with a very simple one just to get a bit more familiar with how to write tests and also to ensure things are working as intended. First we create a test file `DReddit_spec.js` inside `test` and add a `contract()` block that looks something like this:

<pre class="javascript"><code>contract('DReddit', () => {

});</code></pre>

> Inside this code block we'll be putting dedicated test cases. The `contract()` function can be considered a "grouping" functionality to group tests, if you will. It comes with a couple of characteristics specific to Embark, but generally one can say that `contract()` is not only the equivalent to, but an extension of Mocha's [describe()]() function.

Author: Ha! At this point we should brace for questions on why Embark introduces another alias then :)

I'll adjust this paragraph accordingly!

To check whether our test setup is working, we add a simple test that passes:

<pre class="javascript"><code>contract('DReddit', () => {

  it ('should work', () => {
    assert.ok(true);
  });
});</code></pre>

Running this using Embark's `test` command should result in an output similar to this:

<pre class="shell"><code>❯ embark test

Compiling contracts
  DReddit
    ✓ should work (0ms) - [0 gas]

  1 passing (5s) - [Total: 2210775 gas]

 > All tests passed
</code></pre>

This works great, let's go ahead and test some actual functionality!

## Testing the creation of post

Let's test the core functionality of our application - the creation of posts. For that we need to do a couple of things: We need to somehow get an instance of our `DReddit` Smart Contract in JavaScript, so we can call methods on it to test if they work, and we also need to configure out testing environment so that the right Smart Contract instances are created.

### Requiring Smart Contract instances

> When running tests, Embark adds a couple of custom functions and objects to the global scope, which are useful or even necessary. One of those functions is a custom `require()` that lets us import Smart Contract instances from an Embark specific path.

Reviewer: which are ~~useful or even~~ necessary
We could probably explain why it is necessary or link to the doc that explains it

Author: Hm.. yes, but it's actually only necessary because we're providing those custom paths to import from. But theoretically, we could have `embark test` output compiled artifacts on the disc as well and have users import from there. So the necessity kind of comes with the fact that we introduce this thing in the first place.

Is there any other reason we introduce a custom `require()`?

For example, in order to get an instance of our `DReddit` Smart Contract within the test, we add the following line to our spec file:

> <pre class="javascript"><code>const DReddit = embark.require('Embark/contracts/DReddit');</code></pre>

Reviewer 2: The user can just do `require` without the `embark.`.  Might be more confusing, but you can explain it

Author: I actually explain it a moment earlier, so I'm gonna remove `embark` from there! 👌

Once we have such a hash (no worries, we've got one prepared), we can use Web3's `fromAscii()` utils to convert that hash to bytes and then send it off using our Smart Contract's `createPost()` method. We can then subscribe to the events we're emitting and check its return value like this:

<pre class="javascript"><code>...
const ipfsHash = 'Qmc5gCcjYypU7y28oCALwfSvxCBskLuPKWpK4qpterKC7z';

contract('DReddit', () => {
  ...
  it ('should be able to create a post and receive it via contract event', async () => {

>     let receipt = await DReddit.methods.createPost(web3.utils.fromAscii(ipfsHash)).send();

Reviewer 2: ```suggestion
    const receipt = await DReddit.methods.createPost(web3.utils.fromAscii(ipfsHash)).send();
```

Author: 👍

    const event = receipt.events.NewPost;
    postId = event.returnValues.postId;
    assert.equal(web3.utils.toAscii(event.returnValues.description), ipfsHash);
  });
});</code></pre>

Notice that we're using `async/await` here because Embark's Smart Contract instance methods return promises. The same can be done without promises as well, it's just a syntactical difference at this point. Running `embark test` should result in two passing tests now!

## Testing correctness of data

> Another good test case would be to check if the stored data such as the description bytes, the owner etc.  resolve back to the correct data. Notice that this is slightly different from what we're testing in our previous test - there we're testing the description bytes emitted by the `NewPost` event. To test this we take advantage of the `postId` created in the previous test, which is available globally now, to fetch the stored post. We then perform a similar check as in the previous test. We also want to test if the owner data of the post is correct, but for that we need to get access to the account that created the post in the first place.

Reviewer: Not sure, it feels weird
```the owner etc.  resolve ```
=>
```the owner etc...  resolve ```

Reviewer 2: The way Pascal wrote it is correct as far as I know

Luckily wallet accounts can be easily accessed as they are emitted by Embark's `config()` function. All we have to do is attaching a resolution handler to `config()` and storing the emitted value:

<pre class="javascript"><code>...
let accounts = [];

config({
  contracts: {
    DReddit: {}
  }
}, (err, _accounts) => {
  accounts = _accounts;
});</code></pre>

Having that in place, our next test could look something like this:

<pre class="javascript"><code>contract('DReddit', () => {
  ...
  it ('post should have correct data', async () => {
    const post = await DReddit.methods.posts(postId).call();
    assert.equal(web3.utils.toAscii(post.description), ipfsHash);
    assert.equal(post.owner, accounts[0]);
  });
});</code></pre>

## Testing `canVote()`

Alright, next up let's quickly test if our `canVote()` method works the way as expected. As voting on posts that don't exist should never work, we will simply call `canVote()` on a post id that doesn't exist. This test is pretty straight forward:

> <pre class="javascript"><code>contract('DReddit', () => {

Reviewer 2: IMO, you don't have to put the `contract(...` everywhere. After one or to example, users probably got the gist of it.

Author: Yea, okay will roll back on that one.

  ...
  it('should not be able to vote in an unexisting post', async () => {
    const userCanVote = await DReddit.methods.canVote("123").call();
    assert.equal(userCanVote, false);
  });
});</code></pre>

We also want to make sure that `canVote()` resolves to `true` in case a user can indeed vote a certain post. We can again reuse the `postId` that we've stored earlier:

> <pre class="javascript"><code>it('should be able to vote in a post if account hasn\'t voted before', async () => {

Reviewer 2: ```suggestion
<pre class="javascript"><code>it('should be able to vote in a post if account has not voted before', async () => {
```
Very aesthetic change. Backslashes are ugly

Author: Ha okay

  const userCanVote = await DReddit.methods.canVote(postId).call();
  assert.equal(userCanVote, true);
});</code></pre>

Wonderful, we have 5 passing tests now!

## Testing `vote()`

Of course we want to test whether one of our application's core features works as well. There's certainly different ways to verify whether `vote()` does what it's supposed to do, but for this tutorial we'll simply check whether the owner account of the vote emitted by the `NewVote` event is the same as the account that performed the vote. We can actually take some inspiration from our previous tests:

<pre class="javascript"><code>contract('DReddit', () => {
  ...
  it("should be able to vote in a post", async () => {
    const receipt = await DReddit.methods.vote(postId, 1).send();
    const Vote = receipt.events.NewVote;

>     assert.equal(Vote.returnValues.owner, accounts[0]);

Reviewer: Shall we explain that accounts[0] is the default account used.
Or make it explicit in the `send`?

Author: Yes I'll add a comment about that. Good point!

    } catch (error){
      assert(error.message.search('revert') > -1, 'Revert should happen');
    }
  });
});</code></pre>

This might look a bit confusing first but it's actually pretty straight forward. In case `vote()` fails, we should not reach the `assert.fail()` call but end up in the `catch()` block immediately. If that was not the case, the test would fail. This is a very common pattern when testing negatives.

Okay, one last time we run `embark test` and if the output looks like the following, we're fully covered in terms of tests!

<pre class="shell"><code>❯ embark test
Compiling contracts

  DReddit
    ✓ should work (0ms) - [0 gas]
    ✓ should be able to create a post and receive it via contract event (60ms) - [160689 gas]
    ✓ post should have correct data (18ms) - [0 gas]
    ✓ should not be able to vote in an unexisting post (14ms) - [0 gas]
    ✓ should be able to vote in a post if account hasn't voted before (12ms) - [0 gas]
    ✓ should be able to vote in a post (42ms) - [65115 gas]
    ✓ shouldn't be able to vote twice (37ms) - [22815 gas]

  7 passing (5s) - [Total: 3130955 gas]

 > All tests passed</code></pre>
 

>  Awesome! If you run into any issues, check out the repository with all steps recorded [here](TODO). In the next and last part of this series, we'll be building a front-end for our DReddit app using React. Until then, feel free to add more tests as you like!

Reviewer 2: Nice. Was just gonna recommend putting a link to the repo where all the tests are implemented.

Author: Done!
